### PR TITLE
[WIP] Adds a type-safe StateMachine builder to be used for life cycles.

### PR DESCRIPTION
**Pull request by cromwellian**
Adds a declarative type-safe StateMachine immutable builder DSL to be used for life cycles. This is inspired by various frameworks I've used before in games and protocols (e.g. Fettle). 

Each state may have onEntry, onExit, onError handlers attached, as well as  event->newstate transition mappings.

All legal states are defined, as well as event-driven transitions, with attached event and error handlers, in the DSL. 

a PlantUML or GraphViz dump of the structure can be obtained. 

The machine uses an internal queue to manage any events triggered by handlers. A sample implementation of AbstractArcHost.performParticleLifecycle() has been ported.

I'm still working on how the queue will be processed or drained, and whether I'll use another thread (e.g. like a browser microtasks queue). Additionally, its currently not thread safe (no mutexes around the mutate state field or eventQueue)

Comments welcome.

### CODE REVIEW

**Comment by csilvestrini** [comment]
Added Mick as a replacement reviewer, since he's been working on the particle lifecycle a lot.

I like the idea of having a proper reusable state machine implementation, though to be honest I'm not completely sold on the DSL aspect of it, it seems like it might be harder to read and debug. 

> I'm still working on how the queue will be processed or drained, and whether I'll use another thread (e.g. like a browser microtasks queue). Additionally, its currently not thread safe (no mutexes around the mutate state field or eventQueue)

I think it's ok leaving this single threaded, without mutexes and without a separate processing thread. That seems simpler, and fits better with our proposed threading model going forward.

**Comment by jasonwyatt** [comment]
This seems very cool, I'm curious why so many of the methods are suspending functions though? What reason do they need to suspend? (KDoc explaining under what scenarios they suspend execution would be awesome)

**Comment by mykmartin** [comment]
I tend to agree with Cameron; not sure about the DSL approach. I'm also not sure about how the events fit in. For example: in `ParticleStateMachine`, a `CreateEvent` is triggered that doesn't appear to do anything and the process for `MaxFailedEvent` seems a little convoluted.

**Comment by cromwellian** [comment]
> I like the idea of having a proper reusable state machine implementation, though to be honest I'm not completely sold on the DSL aspect of it, it seems like it might be harder to read and debug.

The DSL idiom is a standard Kotlin idiom for doing this (https://kotlinlang.org/docs/reference/type-safe-builders.html)

You can see Tinder is doing a similar technique here (https://github.com/Tinder/Scarlet/blob/5457f827884269a7e1a7c1c60d057d0ed5fa26b1/scarlet/src/main/java/com/tinder/scarlet/internal/connection/Connection.kt)

Or a similar pattern in Haskell (https://wickstrom.tech/finite-state-machines/2017/11/10/finite-state-machines-part-1-modeling-with-haskell.html)

The alternative is the boilerplatey Java approach used in Fettle, e.g.

```
val stateMachine.newBuilder().addState(Foo).onEntry(...).addTransition(...).build()
```

I'd argue laying out the states and events -> transitions helps readability once you get used to it.

**Comment by cromwellian** [comment]
> I tend to agree with Cameron; not sure about the DSL approach. I'm also not sure about how the events fit in. For example: in `ParticleStateMachine`, a `CreateEvent` is triggered that doesn't appear to do anything and the process for `MaxFailedEvent` seems a little convoluted.

It's an omission. 'CreateEvent' is supposed to have a on(CreateEvent, Created) that creates a valid transition edge from Instantiated -> Created state, after onCreate() is called. 'Created'

The reason it uses an exception is because the current implementation is that onError() clears pending transition events. The exception could be removed just by adding a 'clearPendingEvents()' call and then transitioning to the MaxFailed state. 

Basically, code needs to be able to short-circuit a sequence of transitions if an error occurs.

**Comment by cromwellian** [comment]
> This seems very cool, I'm curious why so many of the methods are suspending functions though? What reason do they need to suspend? (KDoc explaining under what scenarios they suspend execution would be awesome)

Mostly laziness. Since the handle APIs are suspend functions, as are the particle APIs, calling them from handlers means launching coroutines everytime, but once your big PR lands that removes the suspend keywords, all these can go away.

**Comment by cromwellian** [comment]
> > This seems very cool, I'm curious why so many of the methods are suspending functions though? What reason do they need to suspend? (KDoc explaining under what scenarios they suspend execution would be awesome)
> 
> Mostly laziness. Since the handle APIs are suspend functions, as are the particle APIs, calling them from handlers means launching coroutines everytime, but once your big PR lands that removes the suspend keywords, all these can go away.

Actually, I believe if I use inline functions, I might be able to have my cake and eat it too, by letting you use either? (similar to the way many Kotlin stdlib inline functions on collections work?)

**Comment by mykmartin** [comment]
> The DSL idiom is a standard Kotlin idiom for doing this (https://kotlinlang.org/docs/reference/type-safe-builders.html)

Fair enough. Definitely better than the boilerplate.

> I'd argue laying out the states and events -> transitions helps readability once you get used to it.

Yes, that's probably true.

**Comment by jasonwyatt** [comment]
> > > This seems very cool, I'm curious why so many of the methods are suspending functions though? What reason do they need to suspend? (KDoc explaining under what scenarios they suspend execution would be awesome)
> > 
> > 
> > Mostly laziness. Since the handle APIs are suspend functions, as are the particle APIs, calling them from handlers means launching coroutines everytime, but once your big PR lands that removes the suspend keywords, all these can go away.
> 
> Actually, I believe if I use inline functions, I might be able to have my cake and eat it too, by letting you use either? (similar to the way many Kotlin stdlib inline functions on collections work?)

That could work!

File: java/arcs/core/host/ArcState.kt [hunk 1]
```diff
@@ -57,6 +57,7 @@ enum class ParticleState {
     Instantiated,
     /** onCreate() has been successfully called. */
     Created,
+    Ready,
```
Comment (jasonwyatt) [inline java/arcs/core/host/ArcState.kt:60]: Needs KDoc.

File: java/arcs/core/host/ParticleStateMachine.kt [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+enum class ParticleEvent {
+    StartEvent,
+    StopEvent,
+    FailedEvent,
+    MaxFailedEvent,
+    CreateEvent,
+    ReadyEvent
```
Comment (jasonwyatt) [inline java/arcs/core/host/ParticleStateMachine.kt:39]: I don't think we need the `Event` suffix on all of these, since they're already nicely namespaced under `ParticleEvent`

Also, KDoc for the overall Enum class would be good.

File: java/arcs/core/host/ParticleStateMachine.kt [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                on(StartEvent, Instantiated)
+                on(MaxFailedEvent, MaxFailed)
+                on(StopEvent, Failed_NeverStarted)
+            }
+
+            state(MaxFailed) {
+                // There is no escape
+            }
+        }
+    }
```
Comment (jasonwyatt) [inline java/arcs/core/host/ParticleStateMachine.kt:146]: This is a lot to take-in. It would be good to have nice comment blocks for each state explaining what the state means, and what its transitions imply.

File: java/arcs/core/host/ParticleStateMachine.kt [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    private val checkForMaxFailures: suspend (ParticleContext).() -> Unit = {
```
Comment (jasonwyatt) [inline java/arcs/core/host/ParticleStateMachine.kt:148]:
```suggestion
    private val checkForMaxFailures: suspend ParticleContext.() -> Unit = {
```

File: java/arcs/core/host/ParticleStateMachine.kt [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        /**
+         * Move to [ParticleState.Failed] if this particle had previously successfully invoked
+         * [Particle.onCreate()], else move to [ParticleState.Failed_NeverStarted]. Increments
+         * consecutive failure count, and if it reaches maximum, transitions to
+         * [ParticleState.MaxFailed].
+         */
+        private suspend fun markParticleAsFailed(
+            particleContext: ParticleContext,
+            exception: Exception
+        ) = particleContext.run {
```
Comment (jasonwyatt) [inline java/arcs/core/host/ParticleStateMachine.kt:168]: why is `.run` needed?  

If this method needs to return Unit, it's cleaner/more readable to just do:

```kotlin
private suspend fun markParticleAsFailed(
    particleContext: ParticleContext,
    exception: Exception
) {
    particleContext.trigger(FailedEvent)
}
```

File: java/arcs/core/util/StateMachine.kt [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.util
+
+typealias EntryHandler<MACHINE> = suspend (MACHINE).() -> Unit
+typealias ErrorHandler<MACHINE> = suspend (MACHINE).(Exception) -> Unit
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:14]:
```suggestion
typealias EntryHandler<Machine> = suspend Machine.() -> Unit
typealias ErrorHandler<Machine> = suspend Machine.(Exception) -> Unit
```

File: java/arcs/core/util/StateMachine.kt [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+ * @property state the current/starting state of the machine.
+ */
+abstract class StateMachine<STATE, EVENT, MACHINE : StateMachine<STATE, EVENT, MACHINE>>(
+    var state: STATE
+) {
+    private lateinit var states: Map<STATE, StateNode<STATE, EVENT, MACHINE>>
+    private val eventQueue = mutableListOf<EVENT>()
+
+    /** True if there are no pending events to process. */
+    val idle = eventQueue.isEmpty()
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:60]: Are we worried about thread-safety for this?

Might be worth considering Flow/Channel, as opposed to manually handling the queue.

File: java/arcs/core/util/StateMachine.kt [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+ * Loosely inspired by those used in gaming or protocol frameworks, like Fettle
+ * [http://thehiflyer.github.io/Fettle/] but simplified for Arcs needs, and constructed via
+ * idiomatic Kotlin type-safe builders.
+ *
+ * @param STATE usually an enum or sealed class representing possible states
+ * @param EVENT usually an enum or sealed class representing possible event triggered transitions
+ * @param MACHINE the subclass type that extends [StateMachine]
+ * @property state the current/starting state of the machine.
+ */
+abstract class StateMachine<STATE, EVENT, MACHINE : StateMachine<STATE, EVENT, MACHINE>>(
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:53]: Type variables should be CapitalCamelCase.

File: java/arcs/core/util/StateMachine.kt [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    var state: STATE
+) {
+    private lateinit var states: Map<STATE, StateNode<STATE, EVENT, MACHINE>>
+    private val eventQueue = mutableListOf<EVENT>()
+
+    /** True if there are no pending events to process. */
+    val idle = eventQueue.isEmpty()
+
+    internal data class Transition<EVENT, STATE, MACHINE>(
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:62]: Type variables should be CapitalCamelCase

File: java/arcs/core/util/StateMachine.kt [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        val nextState: STATE,
+        val block: suspend (MACHINE).(STATE) -> Unit
+    )
+
+    internal data class StateNode<STATE, EVENT, MACHINE>(
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:67]: Type variables should be CapitalCamelCase

File: java/arcs/core/util/StateMachine.kt [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+) {
+    private lateinit var states: Map<STATE, StateNode<STATE, EVENT, MACHINE>>
+    private val eventQueue = mutableListOf<EVENT>()
+
+    /** True if there are no pending events to process. */
+    val idle = eventQueue.isEmpty()
+
+    internal data class Transition<EVENT, STATE, MACHINE>(
+        val nextState: STATE,
+        val block: suspend (MACHINE).(STATE) -> Unit
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:64]: No need for parens around the receiver.

File: java/arcs/core/util/StateMachine.kt [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    )
+
+    internal data class StateNode<STATE, EVENT, MACHINE>(
+        val entryHandler: EntryHandler<MACHINE>,
+        val exitHandler: EntryHandler<MACHINE>,
+        val errorHandler: ErrorHandler<MACHINE>,
+        val eventHandlers: Map<EVENT, Transition<EVENT, STATE, MACHINE>>
+    )
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:72]: Class definitions should go at the bottom of the containing class declaration.

File: java/arcs/core/util/StateMachine.kt [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    @Suppress("UNCHECKED_CAST")
+    protected suspend fun transitionTo(newState: STATE) {
+        val destination = requireNotNull(states[newState]) {
+            "$newState is not defined in the state machine."
+        }
+
+        state = newState
+        try {
+            destination.entryHandler.invoke(this as MACHINE)
+        } catch (e: Exception) {
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:127]: Any exception? 

A comment explaining why would be good here, and if it's necessary - maybe `Throwable` is better?

File: java/arcs/core/util/StateMachine.kt [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        val stateNode = requireNotNull(states[state])
+        val eventHandler = requireNotNull(stateNode.eventHandlers[event]) {
+            "Illegal state transition $event while in state $state"
+        }
+
+        try {
+            stateNode.exitHandler.invoke(this as MACHINE)
+            eventHandler.block(this, eventHandler.nextState)
+            transitionTo(eventHandler.nextState)
+        } catch (e: Exception) {
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:106]: Any exception? 

A comment explaining why would be good here, and if it's necessary - maybe `Throwable` is better?

File: java/arcs/core/util/StateMachine.kt [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+        inner class StateHandler(val state: STATE) {
+
+            private var entryHandler: EntryHandler<MACHINE>? = null
+            private var exitHandler: EntryHandler<MACHINE>? = null
+            private var errorHandler: ErrorHandler<MACHINE>? = null
+            private val eventHandlers: MutableMap<EVENT, Transition<EVENT, STATE, MACHINE>> =
+                mutableMapOf()
+
+            /** Invoked whenever this state is entered. */
+            fun onEntry(handler: suspend (MACHINE).() -> Unit) {
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:147]: No need for parens around the receiver.

File: java/arcs/core/util/StateMachine.kt [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                require(entryHandler == null) {
+                    "$state already has entry handler. Only one entry handler allowed per state."
+                }
+                entryHandler = handler
+            }
+
+            /** Invoked whenever this state is exited. */
+            fun onExit(handler: suspend (MACHINE).() -> Unit) {
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:155]: No need for parens around the receiver.

File: java/arcs/core/util/StateMachine.kt [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+                require(exitHandler == null) {
+                    "$state already has exit handler. Only one exit handler allowed per state."
+                }
+                exitHandler = handler
+            }
+
+            /** Invoked whenever an exception occurs transitioning to this state. */
+            fun onError(handler: suspend (MACHINE).(Exception) -> Unit) {
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:163]: No need for parens around the receiver.

File: java/arcs/core/util/StateMachine.kt [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+                require(errorHandler == null) {
+                    "$state already has error handler. Only one error handler allowed per state."
+                }
+                errorHandler = handler
+            }
+
+            fun on(
+                event: EVENT,
+                nextState: STATE,
+                handler: suspend (MACHINE).(state: STATE) -> Unit = {}
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:173]: No need for parens around the receiver.

File: java/arcs/core/util/StateMachine.kt [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+                exitHandler ?: {},
+                errorHandler ?: {},
+                eventHandlers
+            )
+        }
+
+        /**
+         * Defines a state in the machine and returns builder to register entry and error handlers.
+         */
+        fun state(newState: STATE, block: (StateHandler).() -> Unit) =
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:193]: No need for parens around the receiver.

File: java/arcs/core/util/StateMachine.kt [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+                builderStates[state] = build()
+            }
+
+        internal fun build() {
+            states = builderStates
+        }
+    }
+
+    /** Creates and applies a type-safe builder for the state machine. */
+    fun declareStateMachine(builder: (StateMachineBuilder).() -> Unit) =
```
Comment (jasonwyatt) [inline java/arcs/core/util/StateMachine.kt:207]: No need for parens around the receiver.

File: javatests/arcs/core/util/StateMachineTest.kt [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+
+    @Test
+    fun testIllegalTransition() = runBlockingTest {
+        val machine = TestMachine(TestEventQueue)
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            machine.triggerImmediate(TestEntryAndExitEvent)
+        }
+    }
+
+
```
Comment (jasonwyatt) [inline javatests/arcs/core/util/StateMachineTest.kt:120]: Bonus whitespace

File: javatests/arcs/core/util/StateMachineTest.kt [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupExit()
+        }
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupError()
+        }
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupTransition()
+        }
+    }
```
Comment (jasonwyatt) [inline javatests/arcs/core/util/StateMachineTest.kt:139]: It'd be more readable and easier to understand failures if these were split into separate test cases, and the `Illegal*()` things were inlined.

File: javatests/arcs/core/util/StateMachineTest.kt [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+
+/** Tests for [StateMachine]. */
+@RunWith(JUnit4::class)
+class StateMachineTest {
```
Comment (jasonwyatt) [inline javatests/arcs/core/util/StateMachineTest.kt:23]: Great tests!
